keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -143,22 +144,29 @@ The CRD `v1` API will be the same as the `v1beta1` but with all changes to the A
 
 ### Integration tests for GA
 
-TODO: complete this list
-
-* [Pruning for CustomResources KEP](https://github.com/kubernetes/enhancements/pull/709)
-  * TODO
-* [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
-  * TODO
-* CRD versioning/conversion (https://github.com/kubernetes/kubernetes/issues/64136):
+* Ensure CRD themselves are v1<->v1beta1 round trippable
+* CRD versioning/conversion integration test coverage:
   * Ensure what is persisted in etcd matches the storage version
   * Set up a CRD, persist some data, changed the version, and ensure the previously persisted data is readable
   * Ensure discovery docs track a CRD through creation, version addition, version removal, and deletion
   * Ensure `spec.served` is respected
+  * Ensure malformed inputs result in sensible errors, not panics, and that the system recovers appropriately
+* See also [Pruning for Custom Resources KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20180731-crd-pruning.md)
+* See also [Vanilla OpenAPI Subset: structural schemas KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190425-structural-openapi.md)
+* See also [Defaulting for Custom Resources KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190426-crd-defaulting.md)
 
 ### e2e tests for GA
 
 * Custom Resources should be readable and writable at all available versions (test for round trip-ability)
-* A Custom Resource watch should terminate if its CustomResourceDefinition is deleted or updated
+* Introduce v1 API testing in addition to v1beta1 tests
+
+### Conformance tests for GA
+
+* Separate out a subset of e2e test cases and nominate them for promotion to Conformance
+  * test case covering the basics of the CRD v1 API
+  * test cases exemplifying CRD versioning both without and with a webhook converter
+  * test case exemplifying CRD pruning and defaulting
+  * test case exemplifying structural schemas
```

Comment:
- CRD pruning and defaulting: we don't have e2e test for them, and e2e test was not planned in Pruning and Defaulting KEPs. cc @sttts 
- Structural schemas: add testcases in openapi publishing and webhook conversion to make sure the features are disabled when non-structural schema is provided

Reply:
@sttts Thoughts on Conformance test coverage for Pruning/Defaulting/Structural Schemas?

Reply:
We have extensive e2e-like tests for apiextensions-apiserver for all of this. I don't additional value to having real e2e tests, other than maybe a smoke test, but without big coverage.

Reply:
@sttts How about for conformance specifically?

Reply:
Webhook conversion support is something we want in conformance. General CRD CRUD support as well. But pruning, defaulting, structural schemas is something deep in the semantics of the API, without external dependencies that could be broken in a cluster implementation. Do we really try to cover all API surface in conformance tests that it works as defined?

As written above: a smoke test which checks that all three CRD features are switch one is a good idea. Just want to avoid that we copy those 179 unit and integration test cases verifying that these features work as specified bit by bit.

Reply:
Sounds good. I'll remove pruning, defaulting and structural schemas from the conformance list here.

Completely agree about the smoke test.